modules/rtm/R/invert.auto.R

```diff
@@ -126,79 +155,49 @@ invert.auto <- function(observed, invert.options,
-  if (conv.check$error) {
-    warning("Could not calculate Gelman diag. Assuming no convergence.")
-    conv.check$converged <- FALSE
-  }
-  if (conv.check$converged) {
-    # Done
-    message("Post-process")
-    out <- postProcess(i.ngibbs, samps.list)
-  } else {
-    # Loop until convergence
-    continue <- TRUE
-    while (continue & i.ngibbs < ngibbs.max) {
-      if (!quiet) 
-        print(sprintf("Running iterations %d to %d", i.ngibbs, i.ngibbs + ngibbs.step))
-      inits <- lapply(samps.list, getLastRow)
-      inputs <- list()
-      for (i in seq_len(nchains)) {
-        inputs[[i]] <- list(runID = runID_list[i],
-                            inits = inits[[i]],
-                            resume = resume[[i]])
-      }
-      invert.options$ngibbs <- ngibbs.step
-      if (parallel) {
-        output.list <- parallel::parLapply(cl, inputs, invert.function)
-      } else {
-        output.list <- list()
-        for (i in seq_along(inputs)) {
-          print(sprintf("Running chain %d of %d", i, nchains))
-          output.list[[i]] <- invert.function(inputs[[i]])
-        }
-      }
-      i.ngibbs <- i.ngibbs + ngibbs.step
-      samps.list.current <- lapply(output.list, "[[", 'results')
-      resume <- lapply(output.list, "[[", 'resume')
 
-      samps.list <- combineChains(samps.list, samps.list.current)
-      if (!is.null(save.samples)) {
-        saveRDS(list(resume = resume, samps.list = samps.list), file = save.samples)
-      }
-      smcmc <- PEcAn.assim.batch::makeMCMCList(samps.list)
-      conv.check <- convergenceCheck(smcmc)
-      
-      # Check for convergence
-      if (conv.check$error) {
-        warning("Could not calculate Gelman diag. Assuming no convergence.")
-        conv.check$converged <- FALSE
-      }
-      if (conv.check$converged) {
-        # Done
-        out <- postProcess(i.ngibbs, samps.list)
-        continue <- FALSE
-      } else {
-        continue <- TRUE
-      }
+  resume <- lapply(output.list, '[[', 'resume')
+  out <- process_output(output.list = output.list, 
+                        iter_conv_check = iter_conv_check,
+                        save.samples = save.samples)
+
+  # Loop until convergence (skipped if finished == TRUE)
+  while (!out$finished & i.ngibbs < ngibbs.max) {
+    if (!quiet) {
+      message(sprintf("Running iterations %d to %d", i.ngibbs,
+                      i.ngibbs + ngibbs.step))
     }
-    if (i.ngibbs > ngibbs.max & continue) {
-      warning("Convergence was not achieved, and max iterations exceeded. Returning results as 'NA'.")
-      out <- list(results = NA, 
-                  samples = PEcAn.assim.batch::makeMCMCList(samps.list))
+    inits <- lapply(out$samples, getLastRow)
+    inputs <- list()
+    for (i in seq_len(nchains)) {
+      inputs[[i]] <- list(runID = runID_list[i],
+                          inits = inits[[i]],
+                          resume = resume[[i]])
     }
+    invert.options$ngibbs <- ngibbs.step
```

Comment:
Good catch. This line is necessary because the initial number of samples `ngibbs.min` can be different from the number of samples between each convergence check after that (`ngibbs.step`), but that doesn't change as the run progresses. In an upcoming commit, I'll move that to right before the `while` loop begins.